vyper/venom/analysis/loop_detection.py

```diff
@@ -0,0 +1,72 @@
+from vyper.utils import OrderedSet
+from vyper.venom.analysis.analysis import IRAnalysis
+from vyper.venom.analysis.cfg import CFGAnalysis
+from vyper.venom.basicblock import IRBasicBlock
+
+
+class LoopDetectionAnalysis(IRAnalysis):
+    """
+    Detects loops and computes basic blocks
+    and the block which is before the loop
+    """
+
+    # key = start of the loop (last bb not in the loop)
+    # value = all the block that loop contains
+    loops: dict[IRBasicBlock, OrderedSet[IRBasicBlock]]
+
+    done: OrderedSet[IRBasicBlock]
+    visited: OrderedSet[IRBasicBlock]
+
+    def analyze(self):
+        self.analyses_cache.request_analysis(CFGAnalysis)
+        self.loops: dict[IRBasicBlock, OrderedSet[IRBasicBlock]] = dict()
+        self.done = OrderedSet()
+        self.visited = OrderedSet()
+        entry = self.function.entry
+        self.dfs(entry)
+
+    def invalidate(self):
+        return super().invalidate()
+
+    def dfs(self, bb: IRBasicBlock, before: IRBasicBlock = None):
```

Comment:
this function does loop detection, why name it `dfs`?

Reply:
style: we use the convention `foo_r()` for recursive functions, and prefix with underscore (i.e. `_foo_r()`) for "private" functions (those that are not called externally)